Naive Bayes Algorithm are used to perform data analysis


In [4]:
import pandas as pd
import numpy as np
df=pd.read_csv("tested.csv")
df.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [5]:
df.drop(['PassengerId','Name','Fare','Cabin','Embarked','Ticket','SibSp'],axis='columns',inplace=True)
df.head()

Survived  Pclass     Sex   Age  Parch
0         0       3    male  34.5      0
1         1       3  female  47.0      0
2         0       2    male  62.0      0
3         0       3    male  27.0      0
4         1       3  female  22.0      1

set survived is a target and drop from the column

In [6]:
target=df.Survived
inputs=df.drop('Survived',axis='columns')

create a dummy to change the number format in the gender column

In [7]:
dummies=pd.get_dummies(inputs.Sex).astype(int)
dummies.head(5)

female  male
0       0     1
1       1     0
2       0     1
3       0     1
4       1     0

These column should added to the existing "input" table

In [8]:
inputs=pd.concat([inputs,dummies],axis='columns')
inputs.head(5)

Pclass     Sex   Age  Parch  female  male
0       3    male  34.5      0       0     1
1       3  female  47.0      0       1     0
2       2    male  62.0      0       0     1
3       3    male  27.0      0       0     1
4       3  female  22.0      1       1     0

if you add some column to existing column using this

inputs=pd.concat([inputs,existing column name],axis='columns')

drop out sex column "inplace=True" used for changes is applied in the orginal data file , without means just apply the changes taken copy of the file

In [9]:
inputs.drop('Sex',axis='columns',inplace=True)
inputs.head()

Pclass   Age  Parch  female  male
0       3  34.5      0       0     1
1       3  47.0      0       1     0
2       2  62.0      0       0     1
3       3  27.0      0       0     1
4       3  22.0      1       1     0

Check if any column values are null

In [10]:
inputs.columns[inputs.isna().any()]

Index(['Age'], dtype='object')

In [11]:
inputs.Age[:10]

0    34.5
1    47.0
2    62.0
3    27.0
4    22.0
5    14.0
6    30.0
7    26.0
8    18.0
9    21.0
Name: Age, dtype: float64

if you found where is null means use this

In [12]:
inputs[inputs.Age.isnull()]

Pclass  Age  Parch  female  male
10        3  NaN      0       0     1
22        1  NaN      0       1     0
29        3  NaN      0       0     1
33        3  NaN      2       1     0
36        3  NaN      0       1     0
..      ...  ...    ...     ...   ...
408       3  NaN      0       1     0
410       3  NaN      0       1     0
413       3  NaN      0       0     1
416       3  NaN      0       0     1
417       3  NaN      1       0     1

[86 rows x 5 columns]

it shows age column so fill the value using "mean()"

In [13]:
inputs.Age=inputs.Age.fillna(inputs.Age.mean())
inputs.head()

Pclass   Age  Parch  female  male
0       3  34.5      0       0     1
1       3  47.0      0       1     0
2       2  62.0      0       0     1
3       3  27.0      0       0     1
4       3  22.0      1       1     0

To perform train and test the data

In [14]:
import pandas as pd
from sklearn.model_selection import train_test_split

split data into train and test

Case 1️⃣ : WITHOUT random_state
train_test_split(data, test_size=0.2)

First time you run:

Train → A, B, C, D

Test → E

Second time you run:

Train → A, C, D, E

Test → B

Third time you run:

Train → B, C, D, E

Test → A

❌ Test data changes every time
❌ Result (accuracy) also changes
❌ You get confused 😵

In [15]:
X_train,X_test,y_train,y_test=train_test_split(inputs,target,test_size=0.2,random_state=42)

check the split sizes

In [16]:
print("Training data size : ", X_train.shape)
print("Testing data size :",X_test.shape)

Training data size :  (334, 5)
Testing data size : (84, 5)


X_train → Input data for learning (334 rows)

y_train → Correct answers for training

X_test → New data for testing (84 rows)

y_test → Correct answers for testing

Import Naive Bayes model

In [17]:
from sklearn.naive_bayes import GaussianNB

In [18]:
model=GaussianNB()

In [19]:
model.fit(X_train,y_train)

GaussianNB()

Make predictions using test data

In [20]:
y_pred=model.predict(X_test)

Model predicts survival for 84 new passengers

y_pred → predicted answers

In [21]:
print(y_pred[:10])
print(y_test.values[:10])

[0 1 0 0 1 0 1 0 1 0]
[0 1 0 0 1 0 1 0 1 0]


First line → Model prediction

Second line → Actual result

Check accuracy (final result)

In [22]:
from sklearn.metrics import accuracy_score

accuracy=accuracy_score(y_test,y_pred)
print("Accuracy :",accuracy)

Accuracy : 1.0
